clang/lib/Driver/ToolChains/Clang.cpp

```diff
@@ -6924,28 +6927,40 @@ void OffloadBundler::ConstructJobMultipleOutputs(
           llvm::sys::path::stem(Input.getFilename()).str() + "-prelink", "o");
     InputFileName = C.addTempFile(C.getArgs().MakeArgString(TmpName));
     LinkArgs.push_back(InputFileName);
-    // Input files consist of fat libraries and the object(s) to be unbundled.
-    for (const auto &I : Inputs)
-      LinkArgs.push_back(I.getFilename());
+    const ToolChain *HTC = C.getSingleOffloadToolChain<Action::OFK_Host>();
+    // Add crt objects
+    LinkArgs.push_back(TCArgs.MakeArgString(HTC->GetFilePath("crt1.o")));
+    LinkArgs.push_back(TCArgs.MakeArgString(HTC->GetFilePath("crti.o")));
     // Add -L<dir> search directories.
     TCArgs.AddAllArgs(LinkArgs, options::OPT_L);
-    for (const auto &A :
-            TCArgs.getAllArgValues(options::OPT_foffload_static_lib_EQ))
-      LinkArgs.push_back(TCArgs.MakeArgString(A));
+
+    // TODO - We can potentially go through the args and add the known linker
+    // pass through args of --whole-archive and --no-whole-archive.  This
+    // would allow for support for usages of -Wl,--whole-archive
+    // -foffload-static-lib=<lib> -Wl,--no-whole-archive
+    // Input files consist of fat libraries and the object(s) to be unbundled.
+    for (const auto &I : Inputs) {
+      if (I.getType() == types::TY_WholeArchive)
+        LinkArgs.push_back("--whole-archive");
+      LinkArgs.push_back(I.getFilename());
+      if (I.getType() == types::TY_WholeArchive)
+        LinkArgs.push_back("--no-whole-archive");
```

Comment:
I may be missing something, but it feels like in case of multiple fat libraries, this would flaw the linker command:
`--whole-archive lib0.a --no-whole-archive --whole-archive lib1.a --no-whole-archive crtn.o`.

Reply:
that is correct unfortunately.  The reason being is that you can intermingle -foffload-static-lib and -foffload-whole-static-lib on a single command line.

Reply:
I guess the logic can be improved to check the next arg and only add as needed.

Reply:
Could a simple check for the regular archive type of `Input` and adding a `--no-whole-archive` only in this case do the trick? As for the system objects, I think some bool could be set to true from within the loop if there's a need to precede these with `--no-whole-archive`

Reply:
Added some improvements in this area.